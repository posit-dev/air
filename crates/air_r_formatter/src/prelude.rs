pub(crate) use crate::{AsFormat as _, FormatNodeRule, RFormatContext, RFormatter};
pub use biome_formatter::prelude::*;
pub use biome_rowan::{AstNode as _, AstNodeList as _, AstSeparatedList as _};
