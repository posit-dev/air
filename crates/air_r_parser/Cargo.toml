[package]
authors.workspace = true
categories.workspace = true
description = "Air R parser"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_parser"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_factory = { workspace = true }
air_r_syntax = { workspace = true }
biome_diagnostics = { workspace = true }
biome_parser = { workspace = true }
biome_rowan = { workspace = true }
biome_unicode_table = { workspace = true }
serde = { workspace = true, features = ["derive"] }
tracing = { workspace = true }
tree-sitter = { workspace = true }
tree-sitter-r = { workspace = true }

[dev-dependencies]
biome_console = { workspace = true }
biome_diagnostics = { workspace = true }
insta = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true

[lints]
workspace = true
