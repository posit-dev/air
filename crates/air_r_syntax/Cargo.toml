[package]
name = "air_r_syntax"
version = "0.0.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "SyntaxKind and common rowan definitions for air_r_parser"
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
publish = false

[dependencies]
biome_rowan = { workspace = true, features = ["serde"] }
serde = { workspace = true, features = ["derive"] }

[lints]
workspace = true
