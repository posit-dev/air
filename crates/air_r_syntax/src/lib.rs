#[macro_use]
mod generated;
mod file_source;
pub mod string_ext;
mod syntax_node;

pub use self::generated::*;
pub use biome_rowan::{TextLen, TextRange, TextSize, TokenAtOffset, TriviaPieceKind, WalkEvent};
pub use file_source::RFileSource;
pub use syntax_node::*;

use biome_rowan::{RawSyntaxKind, SyntaxKind, TokenText};

impl From<u16> for RSyntaxKind {
    fn from(d: u16) -> RSyntaxKind {
        assert!(d <= (RSyntaxKind::__LAST as u16));
        unsafe { std::mem::transmute::<u16, RSyntaxKind>(d) }
    }
}

impl From<RSyntaxKind> for u16 {
    fn from(k: RSyntaxKind) -> u16 {
        k as u16
    }
}

impl biome_rowan::SyntaxKind for RSyntaxKind {
    const TOMBSTONE: Self = RSyntaxKind::TOMBSTONE;
    const EOF: Self = RSyntaxKind::EOF;

    fn is_bogus(&self) -> bool {
        match self {
            RSyntaxKind::R_BOGUS
            | RSyntaxKind::R_BOGUS_VALUE
            | RSyntaxKind::R_BOGUS_EXPRESSION
            | RSyntaxKind::R_BOGUS_PARAMETER
            | RSyntaxKind::R_BOGUS_ARGUMENT => true,

            RSyntaxKind::TOMBSTONE
            | RSyntaxKind::EOF
            | RSyntaxKind::UNICODE_BOM
            | RSyntaxKind::SEMICOLON
            | RSyntaxKind::COMMA
            | RSyntaxKind::L_CURLY
            | RSyntaxKind::R_CURLY
            | RSyntaxKind::L_BRACK
            | RSyntaxKind::R_BRACK
            | RSyntaxKind::L_PAREN
            | RSyntaxKind::R_PAREN
            | RSyntaxKind::PLUS
            | RSyntaxKind::EQUAL
            | RSyntaxKind::DOTS
            | RSyntaxKind::FUNCTION_KW
            | RSyntaxKind::FOR_KW
            | RSyntaxKind::IN_KW
            | RSyntaxKind::WHILE_KW
            | RSyntaxKind::REPEAT_KW
            | RSyntaxKind::IF_KW
            | RSyntaxKind::ELSE_KW
            | RSyntaxKind::R_INTEGER_LITERAL
            | RSyntaxKind::R_DOUBLE_LITERAL
            | RSyntaxKind::R_COMPLEX_LITERAL
            | RSyntaxKind::R_STRING_LITERAL
            | RSyntaxKind::R_LOGICAL_LITERAL
            | RSyntaxKind::R_NULL_LITERAL
            | RSyntaxKind::NEWLINE
            | RSyntaxKind::WHITESPACE
            | RSyntaxKind::IDENT
            | RSyntaxKind::COMMENT
            | RSyntaxKind::BACKSLASH
            | RSyntaxKind::R_ROOT
            | RSyntaxKind::R_DOTS
            | RSyntaxKind::R_IDENTIFIER
            | RSyntaxKind::R_BINARY_EXPRESSION
            | RSyntaxKind::R_FUNCTION_DEFINITION
            | RSyntaxKind::R_PARAMETERS
            | RSyntaxKind::R_PARAMETER_LIST
            | RSyntaxKind::R_DOTS_PARAMETER
            | RSyntaxKind::R_IDENTIFIER_PARAMETER
            | RSyntaxKind::R_DEFAULT_PARAMETER
            | RSyntaxKind::R_IF_STATEMENT
            | RSyntaxKind::R_ELSE_CLAUSE
            | RSyntaxKind::R_FOR_STATEMENT
            | RSyntaxKind::R_WHILE_STATEMENT
            | RSyntaxKind::R_REPEAT_STATEMENT
            | RSyntaxKind::R_BRACED_EXPRESSIONS
            | RSyntaxKind::R_PARENTHESIZED_EXPRESSION
            | RSyntaxKind::R_CALL
            | RSyntaxKind::R_CALL_ARGUMENTS
            | RSyntaxKind::R_ARGUMENT_LIST
            | RSyntaxKind::R_NAMED_ARGUMENT
            | RSyntaxKind::R_UNNAMED_ARGUMENT
            | RSyntaxKind::R_DOTS_ARGUMENT
            | RSyntaxKind::R_HOLE_ARGUMENT
            | RSyntaxKind::R_EXPRESSION_LIST
            | RSyntaxKind::R_INTEGER_VALUE
            | RSyntaxKind::R_DOUBLE_VALUE
            | RSyntaxKind::R_COMPLEX_VALUE
            | RSyntaxKind::R_STRING_VALUE
            | RSyntaxKind::R_LOGICAL_VALUE
            | RSyntaxKind::R_NULL_VALUE
            | RSyntaxKind::__LAST => false,
        }
    }

    fn to_bogus(&self) -> Self {
        match self {
            // Bogus value
            RSyntaxKind::R_INTEGER_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_DOUBLE_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_COMPLEX_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_STRING_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_LOGICAL_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_NULL_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_BOGUS_VALUE => RSyntaxKind::R_BOGUS_VALUE,

            // Bogus parameter
            RSyntaxKind::R_DOTS_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_IDENTIFIER_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_DEFAULT_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_BOGUS_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,

            // Bogus expression
            RSyntaxKind::R_BINARY_EXPRESSION => RSyntaxKind::R_BOGUS_EXPRESSION,
            RSyntaxKind::R_BOGUS_EXPRESSION => RSyntaxKind::R_BOGUS_EXPRESSION,

            // Bogus argument
            RSyntaxKind::R_NAMED_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_UNNAMED_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_DOTS_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_HOLE_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_BOGUS_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,

            // Bogus
            RSyntaxKind::TOMBSTONE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::EOF => RSyntaxKind::R_BOGUS,
            RSyntaxKind::UNICODE_BOM => RSyntaxKind::R_BOGUS,
            RSyntaxKind::SEMICOLON => RSyntaxKind::R_BOGUS,
            RSyntaxKind::COMMA => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_CURLY => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CURLY => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_BRACK => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BRACK => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_PAREN => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PAREN => RSyntaxKind::R_BOGUS,
            RSyntaxKind::PLUS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::EQUAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::DOTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::FUNCTION_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::FOR_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IN_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::WHILE_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::REPEAT_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IF_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::ELSE_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_INTEGER_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_DOUBLE_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_COMPLEX_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_STRING_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_LOGICAL_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_NULL_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::NEWLINE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::WHITESPACE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IDENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::COMMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::BACKSLASH => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ROOT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_IDENTIFIER => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_DOTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_FUNCTION_DEFINITION => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARAMETERS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARAMETER_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_IF_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ELSE_CLAUSE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_FOR_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_WHILE_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_REPEAT_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BRACED_EXPRESSIONS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARENTHESIZED_EXPRESSION => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CALL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CALL_ARGUMENTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ARGUMENT_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_EXPRESSION_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::__LAST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BOGUS => RSyntaxKind::R_BOGUS,
        }
    }

    #[inline]
    fn to_raw(&self) -> RawSyntaxKind {
        RawSyntaxKind(*self as u16)
    }

    #[inline]
    fn from_raw(raw: RawSyntaxKind) -> Self {
        Self::from(raw.0)
    }

    fn is_root(&self) -> bool {
        matches!(self, RSyntaxKind::R_ROOT)
    }

    fn is_list(&self) -> bool {
        RSyntaxKind::is_list(*self)
    }

    fn is_trivia(self) -> bool {
        matches!(
            self,
            RSyntaxKind::NEWLINE | RSyntaxKind::WHITESPACE | RSyntaxKind::COMMENT
        )
    }

    fn to_string(&self) -> Option<&'static str> {
        RSyntaxKind::to_string(self)
    }
}

impl TryFrom<RSyntaxKind> for TriviaPieceKind {
    type Error = ();

    fn try_from(value: RSyntaxKind) -> Result<Self, Self::Error> {
        if value.is_trivia() {
            match value {
                RSyntaxKind::NEWLINE => Ok(TriviaPieceKind::Newline),
                RSyntaxKind::WHITESPACE => Ok(TriviaPieceKind::Whitespace),
                RSyntaxKind::COMMENT => Ok(TriviaPieceKind::SingleLineComment),
                _ => unreachable!("Not Trivia"),
            }
        } else {
            Err(())
        }
    }
}

/// Text of `token`, excluding all trivia and removing quotes if `token` is a string literal.
pub fn inner_string_text(token: &RSyntaxToken) -> TokenText {
    let mut text = token.token_text_trimmed();
    if token.kind() == RSyntaxKind::R_STRING_VALUE {
        // remove string delimiters
        // SAFETY: string literal token have a delimiters at the start and the end of the string
        let range = TextRange::new(1.into(), text.len() - TextSize::from(1));
        text = text.slice(range);
    }
    text
}
