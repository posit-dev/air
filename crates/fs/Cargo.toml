
[package]
authors.workspace = true
categories.workspace = true
description = "Air filesystem utilities"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "fs"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
path-absolutize = { workspace = true, features = [
    "once_cell_cache",
    "use_unix_paths_on_wasm",
] }

[lints]
workspace = true
