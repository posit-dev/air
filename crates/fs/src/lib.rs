use path_absolutize::Absolutize;
use std::ffi::OsStr;
use std::path::Path;
use std::path::PathBuf;

pub fn has_r_extension(path: &Path) -> bool {
    path.extension()
        .and_then(OsStr::to_str)
        .map_or(false, is_r_extension)
}

pub fn is_r_extension(extension: &str) -> bool {
    matches!(extension, "r" | "R")
}

/// Convert any path to an absolute path (based on the current working
/// directory).
pub fn normalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path = path.as_ref();
    if let Ok(path) = path.absolutize() {
        path.to_path_buf()
    } else {
        path.to_path_buf()
    }
}

/// Convert an absolute path to be relative to the current working directory.
pub fn relativize_path<P: AsRef<Path>>(path: P) -> String {
    let path = path.as_ref();

    #[cfg(target_arch = "wasm32")]
    let cwd = Path::new(".");
    #[cfg(not(target_arch = "wasm32"))]
    let cwd = path_absolutize::path_dedot::CWD.as_path();

    if let Ok(path) = path.strip_prefix(cwd) {
        return format!("{}", path.display());
    }
    format!("{}", path.display())
}
