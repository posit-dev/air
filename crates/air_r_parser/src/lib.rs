mod error;
mod options;
mod parse;

#[allow(unused)]
mod treesitter;

use air_r_factory::RSyntaxFactory;
pub use error::ParseError;
pub use options::RParserOptions;
pub use parse::parse;
pub use parse::parse_r_with_cache;
pub use parse::Parse;

use air_r_syntax::RLanguage;
use biome_parser::tree_sink::LosslessTreeSink;

pub(crate) type RLosslessTreeSink<'source> = LosslessTreeSink<'source, RLanguage, RSyntaxFactory>;
