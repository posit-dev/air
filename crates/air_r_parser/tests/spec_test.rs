use air_r_parser::ParseError;
use air_r_parser::{parse, RParserOptions};
use air_r_syntax::{RRoot, RSyntaxNode};
use biome_console::fmt::{Formatter, Termcolor};
use biome_console::markup;
use biome_diagnostics::display::PrintDiagnostic;
use biome_diagnostics::termcolor;
use biome_diagnostics::DiagnosticExt;
use biome_parser::prelude::ParseDiagnostic;
use biome_rowan::SyntaxNode;
use biome_rowan::SyntaxSlot;
use biome_rowan::{AstNode, SyntaxKind};
use std::fmt::Write;
use std::fs;
use std::path::Path;

#[derive(Copy, Clone)]
pub enum ExpectedOutcome {
    Pass,
    Fail,
    Undefined,
}

#[ignore]
#[test]
fn quick_test() {
    let code = "
    1$foo
    1$2";
    let options = RParserOptions::default();
    let parsed = parse(code, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    println!("{root:#?}");
}

pub fn run(test_case: &str, _snapshot_name: &str, test_directory: &str, outcome_str: &str) {
    let outcome = match outcome_str {
        "ok" => ExpectedOutcome::Pass,
        "error" => ExpectedOutcome::Fail,
        "undefined" => ExpectedOutcome::Undefined,
        _ => panic!("Invalid expected outcome {outcome_str}"),
    };

    let test_case_path = Path::new(test_case);

    let file_name = test_case_path
        .file_name()
        .expect("Expected test to have a file name")
        .to_str()
        .expect("File name to be valid UTF8");

    let content = fs::read_to_string(test_case_path)
        .expect("Expected test path to be a readable file in UTF8 encoding");

    // Normalize to Unix line endings
    let content = line_ending::normalize(content);

    let options = RParserOptions::default();
    let parsed = parse(&content, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    let formatted_ast = format!("{:#?}", root);

    let mut snapshot = String::new();
    writeln!(snapshot, "\n## Input\n\n```R\n{content}\n```\n\n").unwrap();

    writeln!(
        snapshot,
        r#"## AST

```
{formatted_ast}
```

## CST

```
{:#?}
```
"#,
        parsed.syntax()
    )
    .unwrap();

    if parsed.has_errors() {
        let diagnostics: Vec<ParseDiagnostic> = parsed
            .errors()
            .iter()
            .map(ParseError::diagnostic)
            .map(ParseDiagnostic::clone)
            .collect();

        let mut diagnostics_buffer = termcolor::Buffer::no_color();

        let termcolor = &mut Termcolor(&mut diagnostics_buffer);
        let mut formatter = Formatter::new(termcolor);

        for diagnostic in diagnostics {
            let error = diagnostic
                .clone()
                .with_file_path(file_name)
                .with_file_source_code(&content);

            formatter
                .write_markup(markup! {
                    {PrintDiagnostic::verbose(&error)}
                })
                .expect("failed to emit diagnostic");
        }

        let formatted_diagnostics =
            std::str::from_utf8(diagnostics_buffer.as_slice()).expect("non utf8 in error buffer");

        if matches!(outcome, ExpectedOutcome::Pass) {
            panic!("Expected no errors to be present in a test case that is expected to pass but the following diagnostics are present:\n{formatted_diagnostics}")
        }

        writeln!(snapshot, "## Diagnostics\n\n```").unwrap();
        snapshot.write_str(formatted_diagnostics).unwrap();

        writeln!(snapshot, "```\n").unwrap();
    }

    match outcome {
        ExpectedOutcome::Pass => {
            let missing_required = formatted_ast.contains("missing (required)");
            if missing_required
                || parsed
                    .syntax()
                    .descendants()
                    .any(|node: RSyntaxNode| node.kind().is_bogus())
            {
                panic!("Parsed tree of a 'OK' test case should not contain any missing required children or bogus nodes: \n {formatted_ast:#?} \n\n {formatted_ast}");
            }

            let syntax = parsed.syntax();
            if has_bogus_nodes_or_empty_slots(&syntax) {
                panic!("modified tree has bogus nodes or empty slots:\n{syntax:#?} \n\n {syntax}")
            }
        }
        ExpectedOutcome::Fail => {
            if !parsed.has_errors() {
                panic!("Failing test must have diagnostics");
            }
        }
        _ => {}
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => &test_directory,
    }, {
        insta::assert_snapshot!(file_name, snapshot);
    });
}

/// This check is used in the parser test to ensure it doesn't emit
/// bogus nodes without diagnostics.
pub fn has_bogus_nodes_or_empty_slots<L: biome_rowan::Language>(node: &SyntaxNode<L>) -> bool {
    node.descendants().any(|descendant| {
        let kind = descendant.kind();
        if kind.is_bogus() {
            return true;
        }

        if kind.is_list() {
            return descendant
                .slots()
                .any(|slot| matches!(slot, SyntaxSlot::Empty { .. }));
        }

        false
    })
}
