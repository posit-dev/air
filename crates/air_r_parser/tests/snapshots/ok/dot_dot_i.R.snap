---
source: crates/air_r_parser/tests/spec_test.rs
expression: snapshot
---
## Input

```R
..1
..2
..10

function(...) {
  ..1 + ..2
}

# Parses as a function
..1()

```


## AST

```
RRoot {
    bom_token: missing (optional),
    expressions: RExpressionList [
        RDotDotI {
            value_token: DOTDOTI@0..3 "..1" [] [],
        },
        RDotDotI {
            value_token: DOTDOTI@3..7 "..2" [Newline("\n")] [],
        },
        RDotDotI {
            value_token: DOTDOTI@7..12 "..10" [Newline("\n")] [],
        },
        RFunctionDefinition {
            name: FUNCTION_KW@12..22 "function" [Newline("\n"), Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@22..23 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RDots {
                            value_token: DOTS@23..26 "..." [] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@26..27 ")" [] [],
            },
            body: RBracedExpressions {
                l_curly_token: L_CURLY@27..29 "{" [Whitespace(" ")] [],
                expressions: RExpressionList [
                    RBinaryExpression {
                        left: RDotDotI {
                            value_token: DOTDOTI@29..35 "..1" [Newline("\n"), Whitespace("  ")] [],
                        },
                        operator: PLUS@35..37 "+" [Whitespace(" ")] [],
                        right: RDotDotI {
                            value_token: DOTDOTI@37..41 "..2" [Whitespace(" ")] [],
                        },
                    },
                ],
                r_curly_token: R_CURLY@41..43 "}" [Newline("\n")] [],
            },
        },
        RCall {
            function: RDotDotI {
                value_token: DOTDOTI@43..71 "..1" [Newline("\n"), Newline("\n"), Comments("# Parses as a function"), Newline("\n")] [],
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@71..72 "(" [] [],
                items: RArgumentList [],
                r_paren_token: R_PAREN@72..73 ")" [] [],
            },
        },
    ],
    eof_token: EOF@73..74 "" [Newline("\n")] [],
}
```

## CST

```
0: R_ROOT@0..74
  0: (empty)
  1: R_EXPRESSION_LIST@0..73
    0: R_DOT_DOT_I@0..3
      0: DOTDOTI@0..3 "..1" [] []
    1: R_DOT_DOT_I@3..7
      0: DOTDOTI@3..7 "..2" [Newline("\n")] []
    2: R_DOT_DOT_I@7..12
      0: DOTDOTI@7..12 "..10" [Newline("\n")] []
    3: R_FUNCTION_DEFINITION@12..43
      0: FUNCTION_KW@12..22 "function" [Newline("\n"), Newline("\n")] []
      1: R_PARAMETERS@22..27
        0: L_PAREN@22..23 "(" [] []
        1: R_PARAMETER_LIST@23..26
          0: R_PARAMETER@23..26
            0: R_DOTS@23..26
              0: DOTS@23..26 "..." [] []
            1: (empty)
        2: R_PAREN@26..27 ")" [] []
      2: R_BRACED_EXPRESSIONS@27..43
        0: L_CURLY@27..29 "{" [Whitespace(" ")] []
        1: R_EXPRESSION_LIST@29..41
          0: R_BINARY_EXPRESSION@29..41
            0: R_DOT_DOT_I@29..35
              0: DOTDOTI@29..35 "..1" [Newline("\n"), Whitespace("  ")] []
            1: PLUS@35..37 "+" [Whitespace(" ")] []
            2: R_DOT_DOT_I@37..41
              0: DOTDOTI@37..41 "..2" [Whitespace(" ")] []
        2: R_CURLY@41..43 "}" [Newline("\n")] []
    4: R_CALL@43..73
      0: R_DOT_DOT_I@43..71
        0: DOTDOTI@43..71 "..1" [Newline("\n"), Newline("\n"), Comments("# Parses as a function"), Newline("\n")] []
      1: R_CALL_ARGUMENTS@71..73
        0: L_PAREN@71..72 "(" [] []
        1: R_ARGUMENT_LIST@72..72
        2: R_PAREN@72..73 ")" [] []
  2: EOF@73..74 "" [Newline("\n")] []

```
